src/cache.rs

```diff
@@ -123,15 +124,57 @@ pub extern "C" fn load_wasm(
 
 fn do_load_wasm(
     cache: &mut Cache<GoApi, GoStorage, GoQuerier>,
-    contract_checksum: Buffer,
+    checksum: Buffer,
 ) -> Result<Vec<u8>, Error> {
-    let contract_checksum: Checksum = unsafe { contract_checksum.read() }
+    let checksum: Checksum = unsafe { checksum.read() }
         .ok_or_else(|| Error::empty_arg(CACHE_ARG))?
         .try_into()?;
-    let wasm = cache.load_wasm(&contract_checksum)?;
+    let wasm = cache.load_wasm(&checksum)?;
     Ok(wasm)
 }
 
+#[no_mangle]
+pub extern "C" fn pin(cache: *mut cache_t, checksum: Buffer, err: Option<&mut Buffer>) {
+    let r = match to_cache(cache) {
+        Some(c) => catch_unwind(AssertUnwindSafe(move || do_pin(c, checksum)))
+            .unwrap_or_else(|_| Err(Error::panic())),
+        None => Err(Error::empty_arg(CACHE_ARG)),
+    };
+    let _data = handle_c_error(r, err);
+}
```

Comment:
I guess this is the way to return a potential error to Go land. Strange pattern.

Reply:
cgo seems to look into errno to decide if the call was a success or and error and automatically converts a C function like

```
Buffer ibc_packet_timeout(cache_t *cache,
                          Buffer checksum,
                          Buffer env,
                          Buffer msg,
                          DB db,
                          GoApi api,
                          GoQuerier querier,
                          uint64_t gas_limit,
                          bool print_debug,
                          uint64_t *gas_used,
                          Buffer *err);
```

such that it returns a Go style `value, error` pair:

```go
res, err := C.ibc_packet_timeout(cache.ptr, cs, e, p, db, a, q, cu64(gasLimit), cbool(printDebug), &gasUsed, &errmsg)
```

In the case of an error, we pass the message via a different channel (the out error argument `&errmsg`)

See https://utcc.utoronto.ca/~cks/space/blog/programming/GoCgoErrorReturns